mllib/src/main/scala/org/apache/spark/ml/feature/PowerTransform.scala

```diff
@@ -0,0 +1,561 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import org.apache.commons.math3.analysis._
+import org.apache.commons.math3.optim._
+import org.apache.commons.math3.optim.nonlinear.scalar._
+import org.apache.commons.math3.optim.univariate._
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.annotation.Since
+import org.apache.spark.ml._
+import org.apache.spark.ml.linalg._
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.param.shared._
+import org.apache.spark.ml.util._
+import org.apache.spark.mllib.util.MLUtils
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+
+/**
+ * Params for [[PowerTransform]] and [[PowerTransformModel]].
+ */
+private[feature] trait PowerTransformParams extends Params with HasInputCol with HasOutputCol {
+
+  /**
+   * The model type which is a string (case-sensitive).
+   * Supported options: "yeo-johnson", "box-cox".
+   * (default = yeo-johnson)
+   *
+   * @group param
+   */
+  final val modelType: Param[String] = new Param[String](this, "modelType", "The model type " +
+    "which is a string (case-sensitive). Supported options: yeo-johnson (default), and box-cox.",
+    ParamValidators.inArray[String](PowerTransform.supportedModelTypes))
+
+  /** @group getParam */
+  final def getModelType: String = $(modelType)
+
+  setDefault(modelType -> PowerTransform.YeoJohnson)
+
+  /**
+   * param for number of bins to down-sample the curves in statistics computation.
```

Comment:
Here follows the computation of AUC, if there are too many distinct values in some column, we group them into bins before computation.